### ISSUE TITLE
Kraken API requests frequency exceeds standard rate limit

### ISSUE THREAD

**Issue by wearymanateevedaknotstabooresisting** [provinzio/CoinTaxman#21]
The current request frequency has a too small period, resulting in the excess of the Kraken specified rate limit.

This leads to non-requested _get_price_kraken calls, which return 0 and seem to not be re-requested, thus leading to false calculations afterwards (an may be won't be re-queried in another run, since 0 is saved into the DB? Need confirmation, here).

```warning
2021-03-06 13:38:05,744 price_data   WARNING  Querying trades for #PAIR# at YYYY-mm-dd HH-MM-SS+T (offset=20m): Could not retrieve trades: ['EAPI:Rate limit exceeded']
```

**Comment by provinzio** [comment]
Oh, I thought `raise_for_status` ([here](https://github.com/provinzio/CoinTaxman/blob/main/src/price_data.py#L160)) would catch the teapot/rate limit.

It looks like you are right.

#16 could fix this, if it's implemented for different exchanges (which would include Kraken)

**Comment by wearymanateevedaknotstabooresisting** [comment]
> Calling the public endpoints at a frequency of 1 per second (or less) would remain within the rate limits, but exceeding this frequency could cause the calls to be rate limited. If the rate limits are reached, additional calls would be restricted for a few seconds (or possibly longer if calls continue to be made while the rate limits are active).

-Kraken, [https://support.kraken.com/hc/en-us/articles/206548367-What-are-the-API-rate-limits-#1]

A hot fix would be to increase the random uniform `delay` to a range between [1,2]

**Comment by wearymanateevedaknotstabooresisting** [comment]
> Oh, I thought raise_for_status (here) would catch the teapot/rate limit.
- @provinzio 

Actually, Kraken does still send a status code in the 800s if the rate is limited. `raise_for_status` is only triggered with 400/500s codes for server/client errors (https://requests.readthedocs.io/en/master/user/quickstart/). Therefore, no exception will be raised.

**Comment by provinzio** [comment]
> 
> 
> > Oh, I thought raise_for_status (here) would catch the teapot/rate limit.
> > -@provinzio
> 
> Actually, Kraken does still send a status code in the 800s if the rate is limited. `raise_for_status` is only triggered with 400/500s codes for server/client errors (https://requests.readthedocs.io/en/master/user/quickstart/). Therefore, no exception will be raised.
> 
> Kraken also provides an `error` attribute, which would be filled with `['EAPI:Rate limit exceeded']`. So we can explicitly check for the 803 status code or parse the error attribute.

The code checks for the error attribute. It raises a warning and returns 0 as price, which is resolved unfortunately in this case.
We might want to adjust the error handling.
Hotfixing with an increased delay sounds reasonable for the moment.

**Comment by wearymanateevedaknotstabooresisting** [comment]
I re-run with 1 second as the lower duration value and still got some rate limiting. I re-run with 3 seconds which turned out fine, just took a while ... However, in consequence the tax burden significantly changed from big negative to big positive for an example year.

Therefore, I will test-run again and implement an actual raise of a runtime exception, since the results are heavily affected.

**Comment by provinzio** [comment]
Yeah, missing out prices is never good. We should either recheck 0 prices on the next run or raise an exception if no price could be fetched and abort the evaluation altogether. What's your opinion on this?

**Comment by wearymanateevedaknotstabooresisting** [comment]
Normally, I do not like interrupting a program in any way but instead offering a detailed explanation as an error log message and hinting towards criticality with action needed. But a clear interruption is needed if it is a valid pairs with otherwise valid prices. So I could imagine a dialog about exit, retrying, manual price input or ignoring - ignoring is wrong in any case, imo, so this option is only for manual correction after the run (we should not corrupt the DB for later runs, though).

Regarding the non-existing coins, I assumed in my previous program the following: I traded/did some action with a certain coin, so the coin was indeed valid of the exchange and should at least offer information about it on that exchange. For Kraken, all (even very new coins), are directly listed through the Assets and Pairs endpoints. Staked coins are noted with suffix `.S` so the prices are always retrievable through its asset, e.g. prefix `XTZ` for `XTZ.S` tezos staking, trading pairs, e.g. `XTZEUR`.

So the question emerges if a user should process any coin which does not exist in the market through CoinTaxman at all?

**Comment by wearymanateevedaknotstabooresisting** [comment]
I tried several time intervals, at some point I got the rate limiting, even when the interval was big (> 5 s per request) - this size of time interval is non-feasible, imo). Interestingly, the next request goes just fine. Therefore, I think that simple retries are so much better than bigly increasing the delay time.

**Comment by wearymanateevedaknotstabooresisting** [comment]
Retrying works wonders, at least for Kraken API.
